### ISSUE TITLE
Provider caching does not follow "well-known" discovery specification

### ISSUE THREAD

**Issue by jcrsilva** [gruntwork-io/terragrunt#3283]
## Describe the bug

The terraform [service discovery process](https://developer.hashicorp.com/terraform/internals/remote-service-discovery#discovery-process) for modules and providers states that terraform native services should be discovered using the `/.well-known/terraform.json` document present at a specific host.

This will then give you the base paths for a specific service, for example, a provider registry.

The specific example provided on the website is:

```
{
  "modules.v1": "https://modules.example.com/v1/"
}
```

Which is the path that the hashicorp registry and opentofu use for their modules and providers.

However, one can choose to host these services under a different base path, such as `/foo/bar/v1/...`

The current iteration of the provider cache makes the [assumption this is not possible](https://github.com/gruntwork-io/terragrunt/blob/e5e020932745a734e1929f4854df5b5977da6d26/terraform/cache/handlers/provider_direct.go#L59).

```go
		Path:   path.Join("/v1/providers", provider.Namespace, provider.Name, "versions"),
```

## Steps To Reproduce

Run a tf registry hosting a provider registry in path that does not start with `/v1`. Attempt to enable provider cache. Terragrunt run will fail, as cache server will look for modules at `/v1` directly and get a `404` status.

Possibly adding a test case [here](https://github.com/gruntwork-io/terragrunt/blob/e5e020932745a734e1929f4854df5b5977da6d26/cli/provider_cache_test.go#L38) will be enough to surface the issue.

## Expected behavior

The cache server should respect the base path and be able to download the provider.

## Versions

- Terragrunt version: [v0.62.0](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.62.0)
- OpenTofu/Terraform version: 1.9.2
- Environment details (Ubuntu 20.04, Windows 10, etc.): `alpine/terragrunt` container, [this tag](https://hub.docker.com/layers/alpine/terragrunt/1.9.2/images/sha256-7abd33686bd184e247349d51bcc41c555aadbbcc2f3542acc038b00c6628e8a9?context=explore); plus reproduced with latest master ([e5e0209](https://github.com/gruntwork-io/terragrunt/commit/e5e020932745a734e1929f4854df5b5977da6d26))

## Additional context

I had this surface in a client's internal env, apologies due to the lack of screenshots/etc. In this case the issue was worked around by changing the registry to serve directly from `hostname/v1` (that was possible to do); but this should still be looked into because the registry was working fine with terraform but not the cache.

My suggestion is to grab the base path from the `.well-known/terraform.json` and storing it in the `Provider` model, which then can be referenced to build the URL.

I'm happy to submit a PR, but I'm still kind of unfamiliar with the codebase so it may take a while to get merged.

**Comment by levkohimins** [comment]
Hi @jcrsilva, You are absolutely right! Since this was an experimental functionality, it was implemented as simply as possible. We are gradually working on full functionality. In the near future, we will implement your request as well.

**Comment by jcrsilva** [comment]
@levkohimins thanks for looking into it! It's experimental but it has huge potential.

One question (not sure if it's better to discuss this in the context of another bug), I'm trying to get this to work with the mentioned internal registry, but after progressing it seems that [this variable](https://github.com/gruntwork-io/terragrunt/blob/e5e020932745a734e1929f4854df5b5977da6d26/terraform/cache/models/provider.go#L68) only has the file path `/provider_amd64_yadayada.zip` instead of `https://internal-registry.foo.bar/v1/.../provider_amd64_yadayada.zip` but I can't find for the life of me where this is being set.

For the public terraform registry this works just fine and the full URL is set. This means that when `getter` goes to download the object it just fails.

**Comment by levkohimins** [comment]
Hey @jcrsilva,

> One question (not sure if it's better to discuss this in the context of another bug), I'm trying to get this to work with the mentioned internal registry, but after progressing it seems that [this variable](https://github.com/gruntwork-io/terragrunt/blob/e5e020932745a734e1929f4854df5b5977da6d26/terraform/cache/models/provider.go#L68) only has the file path /provider_amd64_yadayada.zip instead of https://internal-registry.foo.bar/v1/.../provider_amd64_yadayada.zip but I can't find for the life of me where this is being set.

This data is received from your (internal) registry. So you need to figure out why your registry is returning such data.

**Comment by jcrsilva** [comment]
@levkohimins yeah noticed that right now. The cache server receives the request from terraform and reverse proxies it to the registry and returns a modified response.

Just figured out what the issue is: the registry is returning relative paths instead of full URLs. Terraform seems to handle it just fine, but the cache server is under the assumption that all returned paths will be full URLs. 

Seems like it is another bug, want me to file it for tracking?

**Comment by levkohimins** [comment]
@jcrsilva, yes, please create another issue.

**Comment by levkohimins** [comment]
Resolved in [v0.64.1](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.64.1) release.
